File: Dome/dome/notifiers.py
"""
Dome WAF – Alert notifiers.

Sends block/detection events to:
  - Discord webhook  (rich embed)
  - Remote syslog    (UDP/TCP, RFC 5424)
  - Generic webhook  (POST JSON to any URL)

All notifiers are fire-and-forget (asyncio background tasks).
A failure in any notifier never blocks the proxy response.
"""
from __future__ import annotations
import asyncio
import json
import logging
import socket
import time
from datetime import datetime, timezone
from typing import Any

import aiohttp

logger = logging.getLogger("dome.notifiers")

# ── Severity colours for Discord embeds ───────────────────────────────────────
_COLOURS = {
    "BLOCK":  0xE74C3C,   # red
    "LOG":    0xF39C12,   # orange
    "ALLOW":  0x2ECC71,   # green
}

_SYSLOG_SEVERITY = {
    "BLOCK": 2,   # CRIT
    "LOG":   4,   # WARNING
    "ALLOW": 6,   # INFO
    "ERROR": 3,   # ERR
}


# ─────────────────────────────────────────────────────────────────────────────
#  Discord Webhook
# ─────────────────────────────────────────────────────────────────────────────
async def send_discord(
    session: aiohttp.ClientSession,
    webhook_url: str,
    event: dict[str, Any],
    *,
    min_action: str = "BLOCK",
) -> None:
    """
    Post a rich embed to a Discord webhook.
    min_action: only send if action >= this severity (BLOCK | LOG | ALLOW).
    """
    order = {"ALLOW": 0, "LOG": 1, "BLOCK": 2}
    if order.get(event.get("action", ""), 0) < order.get(min_action, 2):
        return

    action   = event.get("action", "?")
    ip       = event.get("client_ip", "?")
    method   = event.get("method", "?")
    path     = event.get("path", "?")
    hits     = event.get("hits", [])
    hit_text = "\n".join(f"• `{h['rule_id']}` – {h['description']}" for h in hits[:10])
    if not hit_text:
        hit_text = "_no rule hits_"

    embed = {
        "title": f"🛡️ Dome WAF — {action}",
        "color": _COLOURS.get(action, 0x95A5A6),
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "fields": [
            {"name": "Action",     "value": f"`{action}`",          "inline": True},
            {"name": "Client IP",  "value": f"`{ip}`",              "inline": True},
            {"name": "Request",    "value": f"`{method} {path}`",   "inline": False},
            {"name": "Rules Hit",  "value": hit_text,               "inline": False},
            {"name": "Categories", "value": "`" + "`, `".join(event.get("categories", [])) + "`" if event.get("categories") else "_none_", "inline": True},
            {"name": "Duration",   "value": f"{event.get('duration_ms', 0):.1f} ms", "inline": True},
        ],
        "footer": {"text": "Dome WAF"},
    }

    payload = {"embeds": [embed], "username": "Dome WAF"}

    try:
        async with session.post(
            webhook_url,
            json=payload,
            timeout=aiohttp.ClientTimeout(total=5),
        ) as resp:
            if resp.status not in (200, 204):
                body = await resp.text()
                logger.warning("Discord webhook returned %s: %s", resp.status, body[:200])
    except Exception as exc:
        logger.warning("Discord webhook error: %s", exc)


# ─────────────────────────────────────────────────────────────────────────────
#  Generic HTTP Webhook (POST JSON)
# ─────────────────────────────────────────────────────────────────────────────
async def send_webhook(
    session: aiohttp.ClientSession,
    url: str,
    event: dict[str, Any],
    *,
    headers: dict[str, str] | None = None,
    min_action: str = "BLOCK",
) -> None:
    """POST the raw event JSON to any HTTP endpoint."""
    order = {"ALLOW": 0, "LOG": 1, "BLOCK": 2}
    if order.get(event.get("action", ""), 0) < order.get(min_action, 2):
        return

    req_headers = {"Content-Type": "application/json"}
    if headers:
        req_headers.update(headers)

    try:
        async with session.post(
            url,
            data=json.dumps(event),
            headers=req_headers,
            timeout=aiohttp.ClientTimeout(total=5),
        ) as resp:
            if resp.status >= 400:
                logger.warning("Webhook %s returned %s", url, resp.status)
    except Exception as exc:
        logger.warning("Webhook error (%s): %s", url, exc)


# ─────────────────────────────────────────────────────────────────────────────
#  Syslog (UDP or TCP, RFC 5424 format)
# ─────────────────────────────────────────────────────────────────────────────
class SyslogNotifier:
    """
    Sends RFC-5424 syslog messages to a remote log server.
    Falls back to UDP if TCP is unavailable.
    """

    FACILITY_LOCAL0 = 16

    def __init__(self, host: str, port: int = 514, proto: str = "udp"):
        self.host  = host
        self.port  = port
        self.proto = proto.lower()
        self._sock: socket.socket | None = None

    def _connect(self) -> None:
        if self._sock:
            return
        if self.proto == "tcp":
            self._sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self._sock.settimeout(3)
            self._sock.connect((self.host, self.port))
        else:
            self._sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

    def _format(self, event: dict[str, Any]) -> bytes:
        severity  = _SYSLOG_SEVERITY.get(event.get("action", ""), 6)
        priority  = self.FACILITY_LOCAL0 * 8 + severity
        timestamp = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z"
        hostname  = socket.gethostname()
        app_name  = "dome-waf"
        msg_id    = event.get("action", "-")
        rules_hit = ",".join(h["rule_id"] for h in event.get("hits", []))
        message   = (
            f"action={event.get('action','-')} "
            f"ip={event.get('client_ip','-')} "
            f"method={event.get('method','-')} "
            f"path={event.get('path','-')} "
            f"rules={rules_hit or '-'} "
            f"status={event.get('status_code','-')}"
        )
        # RFC 5424:  <PRIORITY>VERSION TIMESTAMP HOSTNAME APP-NAME PROCID MSGID MSG
        line = f"<{priority}>1 {timestamp} {hostname} {app_name} - {msg_id} - {message}"
        return line.encode("utf-8")

    def send(self, event: dict[str, Any], *, min_action: str = "BLOCK") -> None:
        order = {"ALLOW": 0, "LOG": 1, "BLOCK": 2}
        if order.get(event.get("action", ""), 0) < order.get(min_action, 2):
            return
        try:
            self._connect()
            data = self._format(event)
            if self.proto == "tcp":
                self._sock.sendall(data + b"\n")
            else:
                self._sock.sendto(data, (self.host, self.port))
        except Exception as exc:
            logger.warning("Syslog send error (%s:%s): %s", self.host, self.port, exc)
            self._sock = None   # force reconnect next time

    # Convenience async wrapper so callers can await it
    async def send_async(self, event: dict[str, Any], *, min_action: str = "BLOCK") -> None:
        loop = asyncio.get_event_loop()
        await loop.run_in_executor(None, lambda: self.send(event, min_action=min_action))


# ─────────────────────────────────────────────────────────────────────────────
#  Notifier Manager  (wires everything together)
# ─────────────────────────────────────────────────────────────────────────────
class NotifierManager:
    """
    Holds all configured notifiers.
    Call  manager.dispatch(session, event)  from the proxy – fire and forget.
    """

    def __init__(self, config: dict):
        nc = config.get("notifications", {})

        self.discord_url    = nc.get("discord_webhook")
        self.discord_min    = nc.get("discord_min_action", "BLOCK")

        self.webhook_url    = nc.get("webhook_url")
        self.webhook_headers = nc.get("webhook_headers", {})
        self.webhook_min    = nc.get("webhook_min_action", "BLOCK")

        syslog_cfg = nc.get("syslog", {})
        if syslog_cfg.get("enabled"):
            self.syslog = SyslogNotifier(
                host=syslog_cfg["host"],
                port=int(syslog_cfg.get("port", 514)),
                proto=syslog_cfg.get("proto", "udp"),
            )
            self.syslog_min = syslog_cfg.get("min_action", "BLOCK")
        else:
            self.syslog     = None
            self.syslog_min = "BLOCK"

    def has_any(self) -> bool:
        return bool(self.discord_url or self.webhook_url or self.syslog)

    def dispatch(
        self,
        session: aiohttp.ClientSession,
        event: dict[str, Any],
    ) -> None:
        """Schedule all notifications as background tasks (non-blocking)."""
        if self.discord_url:
            asyncio.ensure_future(
                send_discord(session, self.discord_url, event, min_action=self.discord_min)
            )
        if self.webhook_url:
            asyncio.ensure_future(
                send_webhook(
                    session, self.webhook_url, event,
                    headers=self.webhook_headers,
                    min_action=self.webhook_min,
                )
            )
        if self.syslog:
            asyncio.ensure_future(
                self.syslog.send_async(event, min_action=self.syslog_min)
            )

File: Dome/dome-rs/src/notifiers.rs
/// Alert notifiers – Discord, generic webhook, syslog (RFC 5424).
/// All are fire-and-forget: spawned as Tokio tasks, never block proxy path.
use reqwest::Client;
use serde_json::{json, Value};
use std::net::UdpSocket;
use tokio::net::TcpStream;
use tokio::io::AsyncWriteExt;
use chrono::Utc;
use crate::config::NotificationsConfig;
use crate::engine::Hit;

fn action_order(a: &str) -> u8 {
    match a { "ALLOW" => 0, "LOG" => 1, "CHALLENGE" => 1, _ => 2 }
}

fn should_send(event_action: &str, min_action: &str) -> bool {
    action_order(event_action) >= action_order(min_action)
}

// ── Discord ───────────────────────────────────────────────────────────────────
pub async fn send_discord(
    client: &Client,
    webhook_url: &str,
    action: &str,
    ip: &str,
    method: &str,
    path: &str,
    hits: &[Hit],
    score: u32,
    min_action: &str,
) {
    if !should_send(action, min_action) { return; }

    let colour: u32 = match action {
        "BLOCK" => 0xE74C3C,
        "LOG" | "CHALLENGE" => 0xF39C12,
        _ => 0x2ECC71,
    };

    let hit_text: String = hits.iter().take(10)
        .map(|h| format!("• `{}` – {}", h.rule_id, h.description))
        .collect::<Vec<_>>()
        .join("\n");
    let hit_text = if hit_text.is_empty() { "_no hits_".into() } else { hit_text };

    let cats: Vec<&str> = {
        let mut seen = std::collections::HashSet::new();
        hits.iter().filter(|h| seen.insert(h.category)).map(|h| h.category).collect()
    };
    let cats_str = if cats.is_empty() {
        "_none_".into()
    } else {
        cats.iter().map(|c| format!("`{c}`")).collect::<Vec<_>>().join(", ")
    };

    let payload = json!({
        "username": "Dome WAF",
        "embeds": [{
            "title": format!("🛡️ Dome WAF — {action}"),
            "color": colour,
            "timestamp": Utc::now().to_rfc3339(),
            "fields": [
                { "name": "Action",     "value": format!("`{action}`"), "inline": true  },
                { "name": "Client IP",  "value": format!("`{ip}`"),     "inline": true  },
                { "name": "Request",    "value": format!("`{method} {path}`"), "inline": false },
                { "name": "Score",      "value": format!("{score} pts"), "inline": true },
                { "name": "Rules Hit",  "value": hit_text,              "inline": false },
                { "name": "Categories", "value": cats_str,              "inline": true  },
            ],
            "footer": { "text": "Dome WAF" }
        }]
    });

    if let Err(e) = client.post(webhook_url).json(&payload).send().await {
        tracing::warn!("Discord webhook error: {e}");
    }
}

// ── Generic HTTP webhook ──────────────────────────────────────────────────────
pub async fn send_webhook(
    client: &Client,
    url: &str,
    extra_headers: &std::collections::HashMap<String, String>,
    event: &Value,
    min_action: &str,
) {
    let action = event["action"].as_str().unwrap_or("");
    if !should_send(action, min_action) { return; }

    let mut req = client.post(url).json(event);
    for (k, v) in extra_headers {
        req = req.header(k.as_str(), v.as_str());
    }
    if let Err(e) = req.send().await {
        tracing::warn!("Webhook error ({url}): {e}");
    }
}

// ── Syslog UDP ────────────────────────────────────────────────────────────────
pub fn send_syslog_udp(
    host: &str,
    port: u16,
    action: &str,
    ip: &str,
    method: &str,
    path: &str,
    hits: &[Hit],
    min_action: &str,
) {
    if !should_send(action, min_action) { return; }

    let severity: u8 = match action { "BLOCK" => 2, "LOG" => 4, _ => 6 };
    let facility: u8 = 16; // local0
    let priority = facility * 8 + severity;
    let ts = Utc::now().format("%Y-%m-%dT%H:%M:%S%.3fZ");
    let hostname = hostname::get()
        .map(|h| h.to_string_lossy().into_owned())
        .unwrap_or_else(|_| "localhost".into());
    let rules: String = hits.iter().map(|h| h.rule_id).collect::<Vec<_>>().join(",");
    let msg = format!(
        "<{priority}>1 {ts} {hostname} dome-waf - {action} - action={action} ip={ip} method={method} path={path} rules={rules}"
    );

    match UdpSocket::bind("0.0.0.0:0") {
        Ok(sock) => {
            let addr = format!("{host}:{port}");
            if let Err(e) = sock.send_to(msg.as_bytes(), &addr) {
                tracing::warn!("Syslog UDP send error: {e}");
            }
        }
        Err(e) => tracing::warn!("Syslog UDP bind error: {e}"),
    }
}

// ── Syslog TCP ────────────────────────────────────────────────────────────────
pub async fn send_syslog_tcp(
    host: &str,
    port: u16,
    action: &str,
    ip: &str,
    method: &str,
    path: &str,
    hits: &[Hit],
    min_action: &str,
) {
    if !should_send(action, min_action) { return; }

    let severity: u8 = match action { "BLOCK" => 2, "LOG" => 4, _ => 6 };
    let priority = 16u8 * 8 + severity;
    let ts = Utc::now().format("%Y-%m-%dT%H:%M:%S%.3fZ");
    let hostname = hostname::get()
        .map(|h| h.to_string_lossy().into_owned())
        .unwrap_or_else(|_| "localhost".into());
    let rules: String = hits.iter().map(|h| h.rule_id).collect::<Vec<_>>().join(",");
    let msg = format!(
        "<{priority}>1 {ts} {hostname} dome-waf - {action} - action={action} ip={ip} method={method} path={path} rules={rules}\n"
    );

    let addr = format!("{host}:{port}");
    match tokio::time::timeout(
        std::time::Duration::from_secs(3),
        TcpStream::connect(&addr),
    ).await {
        Ok(Ok(mut stream)) => {
            let _ = stream.write_all(msg.as_bytes()).await;
        }
        Ok(Err(e)) => tracing::warn!("Syslog TCP connect error ({addr}): {e}"),
        Err(_)     => tracing::warn!("Syslog TCP timeout ({addr})"),
    }
}

// ── Manager ───────────────────────────────────────────────────────────────────
pub struct NotifierManager {
    cfg:    NotificationsConfig,
    client: Client,
}

impl NotifierManager {
    pub fn new(cfg: NotificationsConfig) -> Self {
        let client = Client::builder()
            .timeout(std::time::Duration::from_secs(5))
            .build()
            .expect("reqwest client");
        Self { cfg, client }
    }

    pub fn dispatch(
        &self,
        action: &str,
        ip: &str,
        method: &str,
        path: &str,
        status: u16,
        score: u32,
        hits: &[Hit],
    ) {
        let cfg    = self.cfg.clone();
        let client = self.client.clone();
        let action = action.to_string();
        let ip     = ip.to_string();
        let method = method.to_string();
        let path   = path.to_string();
        let hits_owned: Vec<Hit> = hits.to_vec();

        tokio::spawn(async move {
            // Discord
            if let Some(ref url) = cfg.discord_webhook {
                send_discord(
                    &client, url, &action, &ip, &method, &path,
                    &hits_owned, score, &cfg.discord_min_action,
                ).await;
            }

            // Generic webhook
            if let Some(ref url) = cfg.webhook_url {
                let cats: Vec<_> = {
                    let mut seen = std::collections::HashSet::new();
                    hits_owned.iter().filter(|h| seen.insert(h.category)).map(|h| h.category).collect()
                };
                let event = json!({
                    "product": "dome-waf",
                    "action": action,
                    "client_ip": ip,
                    "method": method,
                    "path": path,
                    "status_code": status,
                    "score": score,
                    "categories": cats,
                    "hits": hits_owned.iter().map(|h| json!({
                        "rule_id": h.rule_id,
                        "description": h.description,
                        "category": h.category,
                        "score": h.score,
                    })).collect::<Vec<_>>(),
                });
                send_webhook(&client, url, &cfg.webhook_headers, &event, &cfg.webhook_min_action).await;
            }

            // Syslog
            if cfg.syslog.enabled {
                if cfg.syslog.proto.to_lowercase() == "tcp" {
                    send_syslog_tcp(
                        &cfg.syslog.host, cfg.syslog.port,
                        &action, &ip, &method, &path,
                        &hits_owned, &cfg.syslog.min_action,
                    ).await;
                } else {
                    send_syslog_udp(
                        &cfg.syslog.host, cfg.syslog.port,
                        &action, &ip, &method, &path,
                        &hits_owned, &cfg.syslog.min_action,
                    );
                }
            }
        });
    }
}
